### DISCUSSION TITLE
Loopback on TCP+SBC Noise

### DISCUSSION THREAD

**Discussion by leelx78** [pschatzmann/arduino-audio-tools#806] [Q&A] [answered]
Hi I'm still new to this but starting to understand the approach. This is working fine: 
I2S -> SBC Encoder -> SBC Decoder -> I2S - Streams defined as follows:
```
EncodedAudioStream decoder(&i2s_stream, new SBCDecoder());
EncodedAudioStream encoder(&decoder, new SBCEncoder(4, 16, 32, SBC_AM_LOUDNESS)); 
StreamCopy copier(encoder, i2s_stream); 

...

void loop() {
  copier.copy();
}
```
Then I included an extra step which sends the stream to a TCP echo server which simply sends all packets back in the same order and I'm trying to to the same through the TCP echo server. 
I2S -> SBC Encoder -> TCP (tx) -> TCP (rx) -> SBC Decoder -> I2S
Streams defined as:

```
WiFiClient client;   
MeasuringStream clientTimed(client);
EncodedAudioStream decoder(&i2s_stream, new SBCDecoder());                    // encode and write
EncodedAudioStream encoder(&clientTimed, new SBCEncoder(4, 16, 32, SBC_AM_LOUDNESS));  // encode and write

StreamCopy copier(encoder, i2s_stream); 
StreamCopy copier_rx(decoder, clientTimed); 

...

void loop() {
  connectIP();  // e.g if client is shut down we try to reconnect
  copier.copy();
  copier_rx.copy();
}
```
and it sounds terrible. I used this method to transfer 24kHz / 16 bit uncompressed audio (>700kbps) (I2S->TCP(tx)->TCP(rx)->I2S) and worked fine so I would think 44.1kHz /. 16 bit + SBC (~350kbps) should be ok?

Am I simply using the copier_rx.copy() wrong?

**Comment by pschatzmann** [comment, marked as answer]
I think this should work, assuming you call the begin method passing the audio info to both, the encoder and the decoder stream.
Can it be that the echo server is dropping some bytes ? This might not only be noticed in your pcm test case when an odd number is lost, but it would have a much bigger impact on the encoded data.

You could also try to decrease the copy size by specifying the length in the constructor of the 2 StreamCopy

**Reply by leelx78** [reply]
Thanks Phil. This is my Python Echo server, very bare:
```
import socket
 
s = socket.socket()         
s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)

s.bind(('0.0.0.0', 8000 ))
s.listen(0)                 
 
while True:
 
    client, addr = s.accept()
 
    while True:
        content = client.recv(2048)
 
        if len(content):
           print("Received %d bytes"%len(content))
           client.send(content)

        
    print("Closing connection")
    client.close()
   ```
   
   are there some headers that are stripped or that I need to watch out for? This is the output from the code:
   
   
   ```
   Received 72 bytes
Received 144 bytes
Received 72 bytes
Received 144 bytes
Received 72 bytes
Received 72 bytes
Received 72 bytes
Received 72 bytes
Received 72 bytes
Received 72 bytes
Received 72 bytes
Received 72 bytes
Received 144 bytes
Received 72 bytes
Received 72 bytes
Received 144 bytes
Received 72 bytes
Received 216 bytes
Received 72 bytes
Received 216 bytes
Received 72 bytes
Received 216 bytes
Received 144 bytes
Received 72 bytes
Received 144 bytes
Received 144 bytes
Received 72 bytes
Received 72 bytes
Received 72 bytes
Received 144 bytes
Received 360 bytes
Received 216 bytes
Received 216 bytes
Received 72 bytes
...
```
I can just hear the audio but it's extremely distorted so definitely something is not lining up here. bit rate seems pretty modest (270kbps or so)

**Reply by pschatzmann** [reply]
Not sure if this makes a difference but you could also try to decrease the copy size by specifying the length in the constructor of the 2 StreamCopy

**Reply by leelx78** [reply]
Another thing I noticed is that if I use these settings, 
```
EncodedAudioStream encoder(&clientTimed, new SBCEncoder(8, 16, 51, SBC_AM_LOUDNESS));  // 48kHz 16 bit encoding high quality (Table 1)
```
The packet size should be 115 according to [https://datatracker.ietf.org/doc/pdf/draft-hoene-avt-rtp-sbc-00.pdf](url) but I always get 114 and multiples back from the echo server, e.g.:


```
void loop() {
  connectIP();  // e.g if client is shut down we try to reconnect
  Serial.printf("Available: TX:%d RX:%d\n", copier.available(), copier_rx.available());
  copier.copy();
  copier_rx.copy();
}

22:32:30.878 -> Available: TX:3072 RX:114
22:32:30.911 -> Available: TX:3072 RX:114
22:32:30.911 -> Available: TX:3072 RX:114
22:32:30.911 -> Available: TX:3072 RX:114
22:32:30.911 -> Available: TX:3072 RX:0
22:32:30.943 -> Available: TX:3072 RX:228
22:32:30.943 -> Available: TX:3072 RX:114
22:32:30.943 -> Available: TX:3072 RX:114
22:32:30.977 -> Available: TX:3072 RX:228
22:32:30.977 -> Available: TX:3072 RX:114
22:32:30.977 -> Available: TX:3072 RX:114
```
Is it possible that the outgoing WiFiClient is trimming one of the SBC packet bytes?

Also, is the buffering inside your Audio streams enough for this or should I create a dedicated queue?

**Comment by pschatzmann** [comment]
I would not rely on what copier.available() is reporting: It's just the avialalbe() of the copy source. For I2S this is alway the max and never the effective value.
You can set the log level to debug to see in detail what's going on

**Reply by leelx78** [reply]
makes sense, this is what the "Info" level is reporting now, no other flags or errors. From what I see the SBC decoder seems happy even if the packet size is not what it should be according to the specification paper for these settings. In fact, even if I use other settings it's always 1 byte less than I expected.

```
09:48:53.427 -> Available: TX:3072 RX:114
09:48:53.427 -> [I] AudioStreams.h : 1098 - ==> Bytes per second: 48000
09:48:53.427 -> [I] StreamCopy.h : 136 - StreamCopy::copy 3072 -> 3072 -> 3072 bytes - in 1 hops
09:48:53.427 -> [I] StreamCopy.h : 136 - StreamCopy::copy 684 -> 684 -> 684 bytes - in 1 hops
09:48:53.460 -> Available: TX:3072 RX:114
09:48:53.460 -> [I] StreamCopy.h : 136 - StreamCopy::copy 3072 -> 3072 -> 3072 bytes - in 1 hops
09:48:53.460 -> [I] StreamCopy.h : 136 - StreamCopy::copy 684 -> 684 -> 684 bytes - in 1 hops
09:48:53.460 -> Available: TX:3072 RX:114
09:48:53.492 -> [I] AudioStreams.h : 1098 - ==> Bytes per second: 63000
09:48:53.492 -> [I] StreamCopy.h : 136 - StreamCopy::copy 3072 -> 3072 -> 3072 bytes - in 1 hops
09:48:53.492 -> [I] StreamCopy.h : 136 - StreamCopy::copy 684 -> 684 -> 684 bytes - in 1 hops
09:48:53.492 -> Available: TX:3072 RX:114
09:48:53.525 -> [I] StreamCopy.h : 136 - StreamCopy::copy 3072 -> 3072 -> 3072 bytes - in 1 hops
```

As for buffering, do I need to use my own queue or do your AudioStream classes already include a, say, 10-50 ms buffer or so?

**Reply by pschatzmann** [reply]
All the encoders and decoders expect a constant stream of data and they will split it up and process a frame as soon as enough data has been received.

Did you try to decrease the copy size ?

If you still need some buffers and is a good choice. Maybe you look e.g. at the RingbufferStream to start with.

**Reply by aqildad-create** [reply]
RingBufferStream (int size=DEFAULT_BUFFER_SIZE)

what is default buffer size and what max size we can give to ring buffer ?

**Reply by pschatzmann** [reply]
Why are you asking me to look this up when you can do it yourself?

**Reply by aqildad-create** [reply]
apology for inconvenience . 

I read RingBufferStream  and other related docs to find default size and maximum range but i failed so that's why i asked here , i was searching for this for last some days .

i know its very annoying for you but these parameter are important for me and i am struggling to find these details .

 i also know your library is well documented and i agree with you these questions are very annoying and disturbing , but i failed to find .

**Reply by pschatzmann** [reply]
Github has a powerful search functionality or you can open the project in Visual Basic Code and do a search there.
Most of the parameters are configured and can be changed in [AudioConfig.h](https://github.com/pschatzmann/arduino-audio-tools/blob/main/src/AudioConfig.h)

**Comment by leelx78** [comment]
Thanks a lot @pschatzmann it's working now!
